specs/lifespan.rst

```diff
@@ -78,6 +82,18 @@ Keys:
 * ``type``: ``lifespan.startup.complete``
 
 
+Startup Failed
+''''''''''''''
+
+Sent by the application when it has failed to complete its startup. If a server
+sees this it should log/print the message provided and then exit.
+
+Keys:
+
+* ``type``: ``lifespan.startup.failed``
+* ``message``: Unicode string (optional, defaults to ``""``).
```

Comment:
Wondering if we should use `"message": str` or `"exception": Exception` here:

* It might make sense to defer to the servers' exception logging.
* Passing through the exception would mean that we'd be able to give exception middleware (eg. Sentry) visibility onto startup/shutdown errors.

?

Reply:
Possibly related to this, possibly not... We've managed so far without an explicit "Here's an exception" message, and instead just raise the exceptions, which works great so far. But I think ASGI will eventually want one, because:

* If a client runs any concurrent-branching code (any protocol type) it may need to communicate 0, 1, or many exceptions up the stack.
* If the client runs any timer-driven or clock-driven events within the context of the `lifespan` task, it may need to communicate exceptions up the stack, without actually raising an exception and terminating the task.

(I don't think an exception message type *neccessarily* needs to be a 3.0 thing, but may be relevant to considering which of `message` vs `exc` is a sensible choice here.)

Reply:
Perhaps the synposis to that oughta be: 1. "Let's use `message` for startup/shutdown failures" and 2. "Let's *plan* on introducing an exception message type (valid on all protocols)" ?

Reply:
The ASGI spec says you're not allowed to send anything outside of basic types (strings, ints, etc.) in messages. We'd need to extend that a lot to allow arbitary Exception classes. I think Message should be enough for now?

And yes, I agree we probably want a generic "this is an exception" message at some point.

Reply:
👍